src/core/room.ts

```diff
@@ -129,53 +138,176 @@ export class DefaultRoom implements Room {
+      }
+      // return the original initAfter promise because in this state the
+      // previous room is still being released
+      return initAfter;
+    };
+
     // Setup features
-    this._messages = new DefaultMessages(roomId, realtime, this._chatApi, realtime.auth.clientId, logger);
+    this._messages = new DefaultMessages(
+      roomId,
+      realtime,
+      this._chatApi,
+      realtime.auth.clientId,
+      logger,
+      initFeaturesAfter,
+    );
+
     const features: ContributesToRoomLifecycle[] = [this._messages];
+
     if (options.presence) {
       this._logger.debug('enabling presence on room', { roomId });
-      this._presence = new DefaultPresence(roomId, options, realtime, realtime.auth.clientId, logger);
+      this._presence = new DefaultPresence(
+        roomId,
+        options,
+        realtime,
+        realtime.auth.clientId,
+        logger,
+        initFeaturesAfter,
+      );
       features.push(this._presence);
     }
 
     if (options.typing) {
       this._logger.debug('enabling typing on room', { roomId });
-      this._typing = new DefaultTyping(roomId, options.typing, realtime, realtime.auth.clientId, logger);
+      this._typing = new DefaultTyping(
+        roomId,
+        options.typing,
+        realtime,
+        realtime.auth.clientId,
+        logger,
+        initFeaturesAfter,
+      );
       features.push(this._typing);
     }
 
     if (options.reactions) {
       this._logger.debug('enabling reactions on room', { roomId });
-      this._reactions = new DefaultRoomReactions(roomId, realtime, realtime.auth.clientId, logger);
+      this._reactions = new DefaultRoomReactions(roomId, realtime, realtime.auth.clientId, logger, initFeaturesAfter);
       features.push(this._reactions);
     }
 
     if (options.occupancy) {
       this._logger.debug('enabling occupancy on room', { roomId });
-      this._occupancy = new DefaultOccupancy(roomId, realtime, this._chatApi, logger);
+      this._occupancy = new DefaultOccupancy(roomId, realtime, this._chatApi, logger, initFeaturesAfter);
       features.push(this._occupancy);
     }
 
-    // Setup lifecycle manager - reverse the features so messages always comes in last
-    this._lifecycleManager = new RoomLifecycleManager(this._status, features.toReversed(), logger, 5000);
-
-    // Setup a finalization function to clean up resources
-    let finalized = false;
-    this._finalizer = async () => {
-      // Cycle the channels in the feature and release them from the realtime client
-      if (finalized) {
-        this._logger.debug('Room.finalizer(); already finalized');
-        return;
+    // Wait for features to finish initializing and then finish initializing
+    // the room. Set _asyncOpsAfter to the promise that waits for the room to
+    // finish initializing. This promise is awaited before performing any async
+    // operations at room level (attach and detach).
+    this._asyncOpsAfter = initFeaturesAfter.then(() => {
```

Comment:
I think this should be moved out of the constructor too please, there's quite a lot going on here :)